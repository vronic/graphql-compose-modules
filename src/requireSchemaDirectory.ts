import fs from 'fs';
import { join, resolve, dirname, basename } from 'path';

interface Options {
  extensions: string[];
  include?: RegExp | ((path: string, filename: string) => boolean);
  exclude?: RegExp | ((path: string, filename: string) => boolean);
}

type RequireAstNodeKinds = 'rootType' | 'dir' | 'file';

interface RequireAstBaseNode {
  kind: RequireAstNodeKinds;
  name: string;
  absPath: string;
}

export interface RequireAstRootTypeNode extends RequireAstBaseNode {
  kind: 'rootType';
  children: {
    [key: string]: RequireAstDirNode | RequireAstFileNode;
  };
}

export interface RequireAstDirNode extends RequireAstBaseNode {
  kind: 'dir';
  children: {
    [key: string]: RequireAstDirNode | RequireAstFileNode;
  };
}

export interface RequireAstFileNode extends RequireAstBaseNode {
  kind: 'file';
  code: {
    default?: any;
  };
}

export interface RequireAstResult {
  query?: RequireAstRootTypeNode;
  mutation?: RequireAstRootTypeNode;
  subscription?: RequireAstRootTypeNode;
}

export const defaultOptions: Options = {
  extensions: ['js', 'ts'],
};

export function requireSchemaDirectory(
  m: NodeModule,
  path?: string,
  options: Options = defaultOptions
): RequireAstResult {
  // if no path was passed in, assume the equivelant of __dirname from caller
  // otherwise, resolve path relative to the equivalent of __dirname
  const schemaPath = !path ? dirname(m.filename) : resolve(dirname(m.filename), path);

  const result = {} as RequireAstResult;

  fs.readdirSync(schemaPath).forEach((filename) => {
    const absPath = join(schemaPath, filename);

    if (fs.statSync(absPath).isDirectory()) {
      const dirName = filename;
      const re = /^(query|mutation|subscription)(\.(.*))?$/i;
      const found = dirName.match(re);
      if (found) {
        const opType = found[1].toLowerCase() as keyof RequireAstResult;
        if (!result[opType])
          result[opType] = {
            kind: 'rootType',
            name: opType,
            absPath,
            children: {},
          } as RequireAstRootTypeNode;

        const subField = found[3]; // any part after dot (eg for `query.me` will be `me`)
        if (subField) {
          result[opType].children[subField] = {
            ...requireSubDirectory(m, absPath, options),
            name: subField,
            absPath,
          };
        } else {
          result[opType].children = requireSubDirectory(m, absPath, options).children;
        }
      }
    }
  });

  return result;
}

export function requireSubDirectory(
  m: NodeModule,
  path: string,
  options: Options = defaultOptions
): RequireAstDirNode {
  const result: RequireAstDirNode = {
    kind: 'dir',
    absPath: resolve(dirname(m.filename), path),
    name: basename(path),
    children: {},
  };

  // setup default options
  Object.keys(defaultOptions).forEach((prop) => {
    if (typeof options[prop] === 'undefined') {
      options[prop] = defaultOptions[prop];
    }
  });

  // get the path of each file in specified directory, append to current tree node, recurse
  fs.readdirSync(path).forEach((filename) => {
    const absPath = join(path, filename);

    const stat = fs.statSync(absPath);
    if (stat.isDirectory()) {
      // this node is a directory; recurse
      if (result.children[filename]) {
        throw new Error(
          `You have a folder and file with same name "${filename}" by the following path ${path}. Please remove one of them.`
        );
      }
      result.children[filename] = requireSubDirectory(m, absPath, options);
    } else if (stat.isFile()) {
      // this node is a file
      if (absPath !== m.filename && checkFileInclusion(absPath, filename, options)) {
        // hash node key shouldn't include file extension
        const moduleName = filename.substring(0, filename.lastIndexOf('.'));
        if (result.children[moduleName]) {
          throw new Error(
            `You have a folder and file with same name "${moduleName}" by the following path ${path}. Please remove one of them.`
          );
        }
        result.children[moduleName] = {
          kind: 'file',
          name: moduleName,
          absPath,
          code: m.require(absPath),
        };
      }
    }
  });

  return result;
}

function checkFileInclusion(absPath: string, filename: string, options: Options): boolean {
  return (
    // verify file has valid extension
    new RegExp('\\.(' + options.extensions.join('|') + ')$', 'i').test(filename) &&
    // if options.include is a RegExp, evaluate it and make sure the path passes
    !(options.include && options.include instanceof RegExp && !options.include.test(absPath)) &&
    // if options.include is a function, evaluate it and make sure the path passes
    !(
      options.include &&
      typeof options.include === 'function' &&
      !options.include(absPath, filename)
    ) &&
    // if options.exclude is a RegExp, evaluate it and make sure the path doesn't pass
    !(options.exclude && options.exclude instanceof RegExp && options.exclude.test(absPath)) &&
    // if options.exclude is a function, evaluate it and make sure the path doesn't pass
    !(
      options.exclude &&
      typeof options.exclude === 'function' &&
      options.exclude(absPath, filename)
    )
  );
}
