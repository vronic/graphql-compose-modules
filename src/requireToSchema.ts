import {
  SchemaComposer,
  ObjectTypeComposer,
  ObjectTypeComposerFieldConfigMapDefinition,
  upperFirst,
  forEachKey,
  schemaComposer,
} from 'graphql-compose';
import {
  RequireAstResult,
  RequireAstRootTypeNode,
  RequireAstDirNode,
  RequireAstFileNode,
} from './requireSchemaDirectory';

export function requireToSchema<TContext = any>(ast: RequireAstResult): SchemaComposer<TContext> {
  const sc = new SchemaComposer<TContext>();

  if (ast.query) populateRoot(sc, 'Query', ast.query);
  if (ast.mutation) populateRoot(sc, 'Mutation', ast.mutation);
  if (ast.subscription) populateRoot(sc, 'Subscription', ast.subscription);

  return sc;
}

function populateRoot(
  sc: SchemaComposer<any>,
  rootName: 'Query' | 'Mutation' | 'Subscription',
  astRootNode: RequireAstRootTypeNode
) {
  const tc = sc[rootName];
  Object.keys(astRootNode.children).forEach((key) => {
    createFields(astRootNode.children[key], rootName, tc);
  });
}

function createFields(
  ast: RequireAstDirNode | RequireAstFileNode,
  suffix: string,
  parent: ObjectTypeComposer
) {
  if (ast.kind === 'file') {
    if (ast.name !== 'index') {
      if (!ast.code.default) throw new Error('File must return field config by default export');

      parent.addNestedFields({
        [ast.name]: ast.code.default,
      });
    }
  } else if (ast.kind === 'dir') {
    let name = ast.name;
    let typeDefs: any = {};

    if (name.indexOf('.') !== -1) {
      const namesArray = name.split('.');

      if (namesArray.length === 1) {
        name = namesArray[0];
      } else {
        name = namesArray.reduce((finalName, current) => {
          return finalName + upperFirst(current);
        }, '');
      }
    }

    if (ast.children['index'] && ast.children['index'].kind === 'file') {
      typeDefs = ast.children['index'].code.default || {};
    }

    if (typeDefs.type) {
      if (typeof typeDefs.type === 'string') {
        typeDefs.type = schemaComposer.createObjectTC(typeDefs.type);
      }
    } else {
      typeDefs.type = schemaComposer.createObjectTC(`${suffix}${upperFirst(name)}`);
    }

    Object.keys(ast.children).forEach((key) => {
      createFields(ast.children[key], name, typeDefs.type);
    });

    parent.addNestedFields({
      [ast.name]: {
        resolve: () => ({}),
        ...typeDefs,
      },
    });
  }
}
