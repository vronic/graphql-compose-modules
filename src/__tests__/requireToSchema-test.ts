import { requireSchemaDirectory } from '../requireSchemaDirectory';
import { requireToSchema } from '../requireToSchema';
import { SchemaComposer } from 'graphql-compose';
import { printSchema } from 'graphql/utilities';

describe('requireToSchema()', () => {
  describe('Schema ../../examples/forTests/schema', () => {
    const ast = requireSchemaDirectory(module, '../../examples/forTests/schema');
    const sc = requireToSchema(ast);

    it('should return schema composer', () => {
      expect(sc).toBeInstanceOf(SchemaComposer);
    });

    it('schema', async () => {
      const printedSchema = await printSchema(sc.buildSchema(), { commentDescriptions: true });
      expect(printedSchema).toMatchSnapshot();
    });
  });
});
