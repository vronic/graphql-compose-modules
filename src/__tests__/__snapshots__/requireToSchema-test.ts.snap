// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requireToSchema() Schema ../../examples/forTests/schema schema 1`] = `
"# Provides default value for input field.
directive @default(value: JSON!) on INPUT_FIELD_DEFINITION

# The \`JSON\` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
scalar JSON

type MeType {
  address: MeTypeAddress
  name: Boolean
}

type MeTypeAddress {
  city: Boolean
  street: Boolean
}

type Mutation {
  logs: MutationLogs
  auth: MutationAuth
  user: MutationUser
}

type MutationAuth {
  login: Boolean
  logout: Boolean
}

type MutationLogs {
  nested: MutationLogsNested
}

type MutationLogsNested {
  list: Boolean
}

type MutationUser {
  create: Boolean
  update: Boolean
}

type Query {
  me: MeType
}
"
`;
